testing/scenario_app/android/app/src/main/java/dev/flutter/scenarios/ExternalTextureFlutterActivity.java

```diff
@@ -0,0 +1,468 @@
+    args.put("texture_id", textureId);
+    args.put("texture_width", SURFACE_WIDTH);
+    args.put("texture_height", SURFACE_HEIGHT);
+  }
+
+  private interface SurfaceRenderer {
+    void attach(Surface surface, CountDownLatch onFirstFrame);
+
+    default boolean canProduceFrames() {
+      return true;
+    }
+
+    void repaint();
+
+    void destroy();
+  }
+
+  private static class CanvasSurfaceRenderer implements SurfaceRenderer {
+    private Surface surface;
+    private CountDownLatch onFirstFrame;
+
+    @Override
+    public void attach(Surface surface, CountDownLatch onFirstFrame) {
+      this.surface = surface;
+      this.onFirstFrame = onFirstFrame;
+    }
+
+    @Override
+    public void repaint() {
+      Canvas canvas =
+          VERSION.SDK_INT >= VERSION_CODES.M
+              ? surface.lockHardwareCanvas()
+              : surface.lockCanvas(null);
+      Paint paint = new Paint();
+      paint.setShader(
+          new LinearGradient(
+              0.0f,
+              0.0f,
+              canvas.getWidth(),
+              canvas.getHeight(),
+              new int[] {
+                // Cyan (#00FFFF)
+                0xFF00FFFF,
+                // Magenta (#FF00FF)
+                0xFFFF00FF,
+                // Yellow (#FFFF00)
+                0xFFFFFF00,
+              },
+              null,
+              TileMode.REPEAT));
+      canvas.drawPaint(paint);
+      surface.unlockCanvasAndPost(canvas);
+
+      if (onFirstFrame != null) {
+        onFirstFrame.countDown();
+        onFirstFrame = null;
+      }
+    }
+
+    @Override
+    public void destroy() {}
+  }
+
+  @RequiresApi(VERSION_CODES.LOLLIPOP)
+  private static class MediaSurfaceRenderer implements SurfaceRenderer {
+    private final Supplier<MediaExtractor> extractorSupplier;
+    private final int rotation;
+    private CountDownLatch onFirstFrame;
+
+    private Surface surface;
+    private MediaExtractor extractor;
+    private MediaFormat format;
+    private Thread decodeThread;
+    private final AtomicBoolean atomicCanProduceFrames = new AtomicBoolean(true);
+
+    private MediaSurfaceRenderer(Supplier<MediaExtractor> extractorSupplier, int rotation) {
+      this.extractorSupplier = extractorSupplier;
+      this.rotation = rotation;
+    }
+
+    @Override
+    public void attach(Surface surface, CountDownLatch onFirstFrame) {
+      this.surface = surface;
+      this.onFirstFrame = onFirstFrame;
+
+      extractor = extractorSupplier.get();
+      format = extractor.getTrackFormat(0);
+
+      // NOTE: MediaFormat.KEY_ROTATION was not available until 23+, but the key is still handled on
+      // API 21+.
+      format.setInteger(MediaFormat.KEY_ROTATION, rotation);
+
+      decodeThread = new Thread(this::decodeThreadMain);
+      decodeThread.start();
+    }
+
+    @Override
+    public boolean canProduceFrames() {
+      return atomicCanProduceFrames.get();
+    }
+
+    private void decodeThreadMain() {
```

Comment:
Is the logic below strictly necessary?  Ideally, the test would be as simple as possible if it can capture the behavior you want to make sure doesn't break.